superduperdb/ext/utils.py

```diff
@@ -1,7 +1,25 @@
+import os
 import typing as t
 
 
 def str_shape(shape: t.Sequence[int]) -> str:
     if not shape:
         raise ValueError('Shape was empty')
     return 'x'.join(str(x) for x in shape)
+
+
+def get_key(key_name: str) -> str:
+    if key_name not in os.environ:
+        raise ValueError(f'{key_name} not set')
+    else:
+        return os.environ[key_name]
+
+
+def format_prompt(X: str, prompt: str, context: t.Optional[t.List[str]] = None) -> str:
+    if '{context}' in prompt:
```

Comment:
Whoa, that was a toughie! :-D. Particularly, I didn't notice that there was no `f` in front of that variable!

The very existence of this function is a bit weird, but it should be something like this maybe?  At least this makes more sense to me has has less indenting.

```
if context is not None:
    if '{context}' not in prompt:
        raise ValueError(f'A context is required for prompt "{prompt}"')
    prompt = prompt.format(context='\n'.join(context))
return prompt + x
```

Reply:
I agree that it's not ideal.

I'm going to keep the original version because the suggested change doesn't catch the condition where `context` is `None` but there is also  a `{context}` in `prompt`.

There will be some further refactoring for #966 and so I will try to address this then.